packages/flutter/lib/src/services/text_editing.dart

```diff
@@ -94,27 +94,40 @@ class TextSelection extends TextRange {
 
   @override
   String toString() {
-    return '${objectRuntimeType(this, 'TextSelection')}(baseOffset: $baseOffset, extentOffset: $extentOffset, affinity: $affinity, isDirectional: $isDirectional)';
+    final String typeName = objectRuntimeType(this, 'TextSelection');
+    if (!isValid) {
+      return '$typeName.invalid';
+    }
+    return isCollapsed
+      ? '$typeName.collapsed(offset: $baseOffset, affinity: $affinity, isDirectional: $isDirectional)'
+      : '$typeName(baseOffset: $baseOffset, extentOffset: $extentOffset, isDirectional: $isDirectional)';
   }
 
   @override
   bool operator ==(Object other) {
     if (identical(this, other))
       return true;
-    return other is TextSelection
-        && other.baseOffset == baseOffset
+    if (other is! TextSelection)
+      return false;
+    if (!isValid) {
+      return !other.isValid;
+    }
+    return other.baseOffset == baseOffset
         && other.extentOffset == extentOffset
-        && other.affinity == affinity
+        && (!isCollapsed || other.affinity == affinity)
```

Comment:
https://master-api.flutter.dev/flutter/services/TextSelection/affinity.html:

> If the text range is collapsed and has more than one visual location (e.g., occurs at a line break), which of the two locations to use when painting the caret.

So it appears we don't have separate text affinities for each endpoint (I wonder why?) and the affinity is only meaningful for collapsed selections.